File: 917-boats-to-save-people/boats-to-save-people.py
class Solution(object):
    def numRescueBoats(self, people, limit):
        """
        :type people: List[int]
        :type limit: int
        :rtype: int
        """
    
        people.sort()  # Sort first!
        l = 0
        r = len(people) - 1
        boats = 0
        
        while l <= r:
            if people[l] + people[r] <= limit:
                l += 1
            r -= 1
            boats += 1
        
        return boats

# time complexity of O(n log n)

File: 917-boats-to-save-people/boats-to-save-people.js
/**
 * @param {number[]} people
 * @param {number} limit
 * @return {number}
 */
var numRescueBoats = function(people, limit) {
    people.sort((a, b) => a - b); // Sort people by weight
    let i = 0;
    let j = people.length - 1;
    let boats = 0;

    while (i <= j) {
        if (people[i] + people[j] <= limit) {
            i++; // lightest person boards with heaviest
        }
        j--; // heaviest person always boards
        boats++; // we used a boat
    }

    return boats;
};